lib/app/modules/smart/token_refresh_sequence.rb

```diff
@@ -28,133 +31,150 @@ class TokenRefreshSequence < SequenceBase
         metadata do
-          id '02'
+          id '01'
           link 'https://tools.ietf.org/html/rfc6749'
           desc %(
-            Server successfully exchanges refresh token at OAuth token endpoint.
-          )
-        end
-
-        oauth2_params = {
-          'grant_type' => 'refresh_token',
-          'refresh_token' => @instance.refresh_token
-        }
-        oauth2_headers = { 'Content-Type' => 'application/x-www-form-urlencoded' }
-
-        if @instance.confidential_client
-          oauth2_headers['Authorization'] = "Basic #{Base64.strict_encode64(@instance.client_id +
-                                                                                ':' +
-                                                                                @instance.client_secret)}"
-        else
-          oauth2_params['client_id'] = @instance.client_id
+            If the request failed verification or is invalid, the authorization server returns an error response.          )
         end
 
-        @token_response = LoggedRestClient.post(@instance.oauth_token_endpoint, oauth2_params, oauth2_headers)
-        assert_response_ok(@token_response)
+        token_response = perform_refresh_request(@instance.client_id, INVALID_REFRESH_TOKEN)
+        assert_response_bad_or_unauthorized token_response
       end
 
-      test 'Data returned from refresh token exchange contains required information encoded in JSON.' do
+      test 'Refresh token exchange fails when provided invalid Client ID.' do
         metadata do
-          id '03'
-          link 'http://www.hl7.org/fhir/smart-app-launch/'
+          id '02'
+          link 'https://tools.ietf.org/html/rfc6749'
           desc %(
-           The EHR authorization server SHALL return a JSON structure that includes an access token or a message indicating that the authorization request has been denied.
-           access_token, token_type, and scope are required. access_token must be Bearer.
-          )
+            If the request failed verification or is invalid, the authorization server returns an error response.          )
         end
 
-        @token_response_headers = @token_response.headers
-        assert_valid_json(@token_response.body)
-        @token_response_body = JSON.parse(@token_response.body)
+        token_response = perform_refresh_request(INVALID_CLIENT_ID, @instance.refresh_token)
+        assert_response_bad_or_unauthorized token_response
+      end
 
-        assert @token_response_body.key?('access_token'), 'Token response did not contain access_token as required'
+      def validate_and_save_refresh_response(token_response)
+        assert_response_ok(token_response)
+        assert_valid_json(token_response.body)
+        token_response_body = JSON.parse(token_response.body)
+
+        assert token_response_body.key?('access_token'), 'Token response did not contain access_token as required'
```

Comment:
Good catch on `expires_in`.

Regarding moving the code, I would agree on moving that up here from a code clarity perspective.  But we do it this way because we often run into situations where systems are mostly correct, but slightly wrong.  And it is frustrating for developers to be blocked from see how conformant they are on steps 2, 3, and 4 if they need to be 100% conformant on step 1 before continuing.  So we check and save the most important part (the bearer token), then look at all the other pieces of information that are required but that we don't actually use.   I'll see if I can make that more clear.

Reply:
Interesting how `expires_in` is required here but just `recommended` when getting your access token the first time...

Reply:
Moved to required as confirmed in https://github.com/siteadmin/inferno/issues/246#issuecomment-514692915

I opened a ticket for us to provide a warning when expires_in is not provided during the launch sequences.